x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/classification/AbstractAucRoc.java

```diff
@@ -0,0 +1,289 @@
+
+        private final double tpr;
+        private final double fpr;
+        private final double threshold;
+
+        AucRocPoint(double tpr, double fpr, double threshold) {
+            this.tpr = tpr;
+            this.fpr = fpr;
+            this.threshold = threshold;
+        }
+
+        private AucRocPoint(StreamInput in) throws IOException {
+            this.tpr = in.readDouble();
+            this.fpr = in.readDouble();
+            this.threshold = in.readDouble();
+        }
+
+        @Override
+        public int compareTo(AucRocPoint o) {
+            return Comparator.comparingDouble((AucRocPoint p) -> p.threshold).reversed()
+                .thenComparing(p -> p.fpr)
+                .thenComparing(p -> p.tpr)
+                .compare(this, o);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeDouble(tpr);
+            out.writeDouble(fpr);
+            out.writeDouble(threshold);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field("tpr", tpr);
+            builder.field("fpr", fpr);
+            builder.field("threshold", threshold);
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            AucRocPoint that = (AucRocPoint) o;
+            return tpr == that.tpr
+                && fpr == that.fpr
+                && threshold == that.threshold;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(tpr, fpr, threshold);
+        }
+
+        @Override
+        public String toString() {
+            return Strings.toString(this);
+        }
+    }
+
+    private static double interpolate(double x, double x1, double y1, double x2, double y2) {
+        return y1 + (x - x1) * (y2 - y1) / (x2 - x1);
+    }
+
+    public static class Result implements EvaluationMetricResult {
+
+        private final double score;
+        private final List<AucRocPoint> curve;
+
+        public Result(double score, List<AucRocPoint> curve) {
+            this.score = score;
+            this.curve = Objects.requireNonNull(curve);
+        }
+
+        public Result(StreamInput in) throws IOException {
+            this.score = in.readDouble();
+            this.curve = in.readList(AucRocPoint::new);
+        }
+
+        public double getScore() {
+            return score;
+        }
+
+        public List<AucRocPoint> getCurve() {
+            return Collections.unmodifiableList(curve);
+        }
+
+        @Override
+        public String getWriteableName() {
+            return registeredMetricName(Classification.NAME, NAME);
+        }
```

Comment:
I understand that removing the abstract auc_roc `NAME` might cause issues here. But the subclasses could supply a concrete version of these results class or the abstract is the thing that supplies the serialization names.